server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -150,6 +175,12 @@ protected void newResponseAsync(
                 + "the cluster state that are too slow for a transport thread"
         );
         assert ThreadPool.assertCurrentThreadPool(ThreadPool.Names.MANAGEMENT);
+
+        if (request.isRemoteStats()) {
```

Comment:
Could we inline `newRemoteResponseAsync ()` and move this branch into the `andThenApply` call below, rather than having two ways out of this method? Seems neater there to me.

Reply:
If we're in the remote branch, then there's no reason to even launch the whole `additionalStats` work so that's why I took the separate branch. I wanted the remote response to be clearly separated from the big real response, so we won't confuse them and won't accidentally pull some unnecessary parts into the small one or omit them from the big one. I can inline the `newRemoteResponseAsync` inside the `if` for sure, I just thought having a separate method makes it more clear what's going on. The main point is that on remote branch, we don't even need to go through the listener route since the code under the listener wouldn't need to be run.

Reply:
Sure, but we complete `additionalStatsListener` either way (as is right and proper) so the `andThenApply` lambda executes inline.

Reply:
My motivation here btw is that odd control flow is a common source of leaks, e.g. failing to complete all the relevant listeners on all code paths. The code as written here is not leaky, but this sort of thing puts risk on future changes. Sometimes funky control flow is the best way to achieve the desired outcome, especially on a hot path, but the closer to straight-line code we can get things the less chance there is of a future bug.

Reply:
OK I inlined it so that everything happens in one path.